implemented insertWithKey

rbobrowicz opened the pull request:
I implemented the insertWithKey function. The implementation is mostly a copy and paste of the regular insert function with a small change to use the provided function to produce the new value instead of simply replacing it. We might want to refactor this to reduce the code duplication.

I never used Criterion before, and Trie and HashMap don't come have a insertWithKey function, so I'm not really sure if I did this part correctly.

Also included two tests. One for when the key is already present and one for when it's missing.

rbobrowicz committed:
implemented insertWithKey

archblob committed:
Implement, test and benchmark findMin and findMax.

--HG--
extra : rebase_source : 7c648ec2765b8df88425c078e1ecb44659d792b9
extra : histedit_source : e016c92022737f29bab58861e0cbabec489eee7c

archblob committed:
Implement deleteMin, deleteMax, deleteFindMin, deleteFindMax

--HG--
extra : rebase_source : f62ee347ea5b6de15c4f288d67827d96677c5b1d
extra : histedit_source : 5a694e0e358baea7f2e2aaf02313fe4e2335a06f

bos commented:
Thanks for the patch!

I think a more promising way to do this is actually to replace `Core.insert` with your `insertWithKey` function, and then write new `Tree.insert` and `Tree.insertWith` functions _in terms of_ `Core.insertWithKey`. For example:

``` haskell
insert k v m = insertWithKey (\_ v' _ -> v') k v m
```

At least in my hopes, GHC's inliner will do the right thing and eliminate the redundant function calls so that there will be no performance hit from defining `insert` in this way, but you should definitely use criterion to compare before and after performance numbers.

archblob committed:
Pull leftmost/rightmost functions to top level.

--HG--
extra : rebase_source : 07174380cc9c7a0a09bf18cd707a44cc60be0313
extra : amend_source : 35afdf933724db2f02ad58919e302aeb4c6d6650
extra : histedit_source : e28d53fdea7cb33c695b84f894723541c214dd6b

bos committed:
A little reformatting

bos committed:
Better tests for deleteFindMin / deleteFindMax

bos committed:
Merge

bos committed:
Merge

rbobrowicz committed:
Merge branch 'master' of https://github.com/rbobrowicz/critbit

Conflicts:
	Data/CritBit/Tree.hs
	tests/Properties.hs

rbobrowicz committed:
reimplemented insert and added insertWith

rbobrowicz commented:
I reimplemented `insert` in terms of `insertWithKey` as well as added `insertWith` (also in terms of `insertWithKey`). This necessitated moving `insert` from Core to Tree to avoid circular dependencies. 

I tried using Criterion to measure the performance of the different implementations, but on my machine the tests generate a large variance making it hard to fully understand the effects of the change. It did however seem to me that using lambdas to implement `insert` and `insertWith` was faster than using `const`. However, maybe someone with a more stable machine can prove otherwise.

P.S. I'm new to using GitHub, so hopefully I'm doing everything correctly, pleas say otherwise if I'm not.

bos commented:
Thanks. I pulled your changes, then did a little reorganising until I ended up with 323cc3ad9ff4f6e4d084a8870ae2536a49c04921.

I squashed your commits together, then moved `insertWithKey` back into `Core`.
